mesonbuild/build.py

```diff
@@ -1191,12 +1191,22 @@ def process_kwargs(self, kwargs, environment):
         # Visual Studio module-definitions file
         if 'vs_module_defs' in kwargs:
             path = kwargs['vs_module_defs']
-            if os.path.isabs(path):
-                self.vs_module_defs = File.from_absolute_file(path)
+            if isinstance(path, str):
+                if os.path.isabs(path):
+                    self.vs_module_defs = File.from_absolute_file(path)
+                else:
+                    self.vs_module_defs = File.from_source_file(environment.source_dir, self.subdir, path)
+                # link_depends can be an absolute path or relative to self.subdir
+                self.link_depends.append(path)
+            elif isinstance(path, File):
+                # When passing a generated file.
+                self.vs_module_defs = path
+                # link_depends can be an absolute path or relative to self.subdir
+                self.link_depends.append(path.absolute_path(environment.source_dir, environment.build_dir))
```

Comment:
I think this is a good opportunity to make `target.link_depends` require File objects instead of strings. We've been slowly moving the codebase to use File objects when referring to files.

You should only need to change `generate_link()` in `backends/ninjabackend.py`.

... which is slightly suspect because it means that the VS backends don't add `link_depends` dependencies in build targets, which should also be fixed. That depends on [another PR](https://github.com/mesonbuild/meson/pull/1596/files#diff-2736199afea34a4f93305938406f439aR502), so I guess I'll handle that once this PR is merged.